src/humanloop/decorators/tool.py

```diff
@@ -0,0 +1,440 @@
+        ]
+    if origin is typing.Union:
+        sub_types = typing.get_args(annotation)
+        if sub_types[-1] is type(None):
+            # Union is an Optional type
+            if len(sub_types) == 2:
+                return tuple(_parse_annotation(sub_types[0]))
+            return (
+                origin,
+                *[_parse_annotation(sub_type) for sub_type in sub_types[:-1]],
+            )
+        # Union type
+        return [
+            origin,
+            *[_parse_annotation(sub_type) for sub_type in sub_types],
+        ]
+
+    raise ValueError(f"Unsupported origin: {origin}")
+
+
+def _annotation_parse_to_json_schema(arg: Union[list, tuple]) -> Mapping[str, Union[str, Mapping, Sequence]]:
+    """
+    Convert parse result from _parse_annotation to JSON Schema for a parameter.
+
+    The function recursively converts the nested type hints to a JSON Schema.
+
+    Note that 'any' is not supported by JSON Schema, so we allow any type as a workaround.
+
+    Examples:
+        [str] -> {"type": "string"}
+        (str) -> {"type": ["string", "null"]}
+
+        [list, [str]] -> {"type": "array", "items": {"type": "string"}}
+        (list, [str]) -> {"type": ["array", "null"], "items": {"type": "string"}}
+
+        [dict, [str], [int]] ->
+            {
+                "type": "object",
+                "properties": {
+                    "key": {"type": "string"},
+                    "value": {"type": "integer"}
+                }
+            }
+
+        [list, [list, str]] ->
+            {
+                "type": "array",
+                "items": {
+                    "type": "array",
+                    "items": {"type": "string"}
+                }
+            }
+
+        tuple[str, int, list[str]] ->
+            {
+                type: "array",
+                items: [
+                    {"type": "string"},
+                    {"type": "integer"},
+                    {
+                        "type": "array",
+                        "items": {"type": "string"}
+                    }
+                ]
+            }
+
+        Union[str, int] ->
+            {
+                "anyOf": [
+                    {"type": "string"},
+                    {"type": "integer"}
+                ]
+            }
+
+        dict[int, list] ->
+            {
+                "type": "object",
+                "properties": {
+                    "key": {"type": "integer"},
+                    "value": {
+                        "type": "array",
+                        "items": {"type": ["string", "integer", "number", "boolean", "object", "array", "null"]}
+                    }
+                }
+            }
+
+        Optional[list] ->
+            {
+                "type": ["array", "null"],
+                "items": {"type": ["string", "integer", "number", "boolean", "object", "array", "null"]},
+            }
+    """
+    is_nullable = isinstance(arg, tuple)
```

Comment:
this is very confusing - what does `is_nullable` mean, and why is it true when arg is a tuple?

Reply:
This is documented in the docstring of _parse_annotation. I wanted to mark a parameter as optional. If an extracted type is packaged in a list, that parameter is mandatory, if it's a tuple then it is optional.

Reply:
I have refactored the two methods that parse annotation to be more readable